examples/protocol_specification_ex/default.yaml

```diff
@@ -24,3 +24,10 @@ ct:ErrorCode: |
     }
   ErrorCodeEnum error_code = 1;
 ...
+---
+reply:
+  bytes: [bytes, error]
+  error: []
+roles: {agent}
+end_states: [open, closed]
```

Comment:
`end_states` here should list the states in which _terminated_ default dialogues can be. 

I don't think terminated default dialogues can be _open_ or _closed_, they are probably always _closed_.

I think we should first see how a default dialogues can terminate. From the reply structure, it appears that a default dialogue is a continues stream of `bytes` messages until a `bytes` message is replied to with an `error` message. In which case the dialogue is then marked as terminated. If so, an `error` message terminates default dialogues. As long as no `error` message, the dialogue is open and agents can send either `bytes` or `error` messages. As soon as an `error` message is sent, the dialogue is marked as terminated, in which case the end_state is always `error_happend` (or something with a better name).

Probably a better name for `end_states` is `outcome`, (also consistent with the literature), which probably better conveys that we are talking about the state of terminated dialogues of X protocol.

Reply:
I start thinking that `bytes` cannot be answered with other `bytes`. In other words, a default dialogue should have only one message in case of bytes, optionally followed by an error message from the other end.

With that change, the possible sequences of messages are:
```
a) bytes
b) bytes, error
```

Can we express such optionality? In this case, dialogue `(a)` would be in a "maybe closed" state.

Shouldn't we also prevent a dialogue from starting with `error`?

Reply:
Can I remind you both that `end_states` simply defines an enum which then can be used by the user in setting `end_states` on `dialogue_stats` once they are reached. Hence, the `end_states` do not get automatically set by the dialogue state. Therefore, all that's required here is to determine sensible states a dialogue can be in. So in this light and @5A11's comment I would suggest we use only `end_states = {closed}`. Note though, that not all dialogues are closed. Only those specified as such by the user. By default they are in an undefined endstate. I disagree that `end_states` can only be reached once an error happens. It is up to the user to define (e.g. based on message content) when the dialogue has ended. This PR is not about changing anything on the dialogue generator/naming of fields/optionality of fields. So any suggestions wrt to altering this should be moved to an issue and discussed there. I don;t see the value in not permitting `bytes` to be responded with `bytes`.

Reply:
`closed` isn't really a good name for it, as per definition, `end_states` are always about "closed" dialogues. If default dialogues always look like a potentially infinite `bytes` stream, optionally at some point an `error` which terminates the dialogue, I think either leave `end_states` empty (signifying there no different end states for this dialogue), or name it `failed`/`erratic`/`error_happened`/... to convey such a dialogue's actual state.

Reply:
So we can go with `[successful, failed]`. The key is, you want to distinguish dialogues which are finished and not failed from those which are still ongoing. So we need two end states.

Reply:
The wider issue here is that it should probably be `states` rather than `end_states` because then we could have `[successful, failed, open]` which is really much more useful.

Reply:
Exactly, the problem is we are dealing with two concepts: whether the dialogue is open/terminated, and what the outcome is once it is terminated. 

For now, we can combine `open/closed` and `end-states` in a single `states` and have `[successful, failed, open]`. Though, based on the definition, a default dialogue can never be in the `successful` state, only `open` or `failed`.